collector/pipelinerun.go

```diff
@@ -126,6 +131,139 @@ func (f *taskRunGapEventFilter) Generic(event.GenericEvent) bool {
 	return false
 }
 
+type ReconcilePVCThrottled struct {
+	client        client.Client
+	scheme        *runtime.Scheme
+	eventRecorder record.EventRecorder
+	prCollector   *ThrottledByPVCQuotaCollector
+	nsCache       map[string]struct{}
+	listMutex     sync.RWMutex
+}
+
+func (r *ReconcilePVCThrottled) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithCancel(ctx)
+	defer cancel()
+	log := log.FromContext(ctx)
+
+	// so we don't collide with the periodic relist/reset in Start; we should still be able to process events concurrently
+	r.listMutex.RLock()
+	defer r.listMutex.RUnlock()
+
+	//TODO remember, keep track of when pipeline-service and RHTAP starts moving from v1beta1 to v1
+	pr := &v1beta1.PipelineRun{}
+	err := r.client.Get(ctx, types.NamespacedName{Namespace: request.Namespace, Name: request.Name}, pr)
+	if err != nil && !errors.IsNotFound(err) {
+		return reconcile.Result{}, err
+	}
+	if err != nil {
+		// given the various complexities around deletion processing and controllers, we do not decrement our
+		// metric in real time, but rather reset the metrics in our background poller attuned to the operator's pruner
+		// interval.
+		log.V(4).Info(fmt.Sprintf("processing deleted pipelinerun %q", request.NamespacedName))
+		return reconcile.Result{}, nil
+	}
+
+	// based on our WithEventFilter we should only be getting called if this got throttled by PVC
+	log.V(4).Info(fmt.Sprintf("recording pvc throttling for %q", request.NamespacedName))
+	r.prCollector.incPVCThrottle(pr)
+	return reconcile.Result{}, nil
+}
+
+// Start - we do a long running runnable to reset the metric in case we miss delete events, as controller relist does not duplicate
+// delete events like it can create/update events
+func (r *ReconcilePVCThrottled) Start(ctx context.Context) error {
+	//this matches the scheduling interval for pruner in the operator's TektonConfig object
```

Comment:
I'm going to accept that for now, but we should not make any assumptions on how long is a PLR TTL once it is completed.

Reply:
agreed